treewide: support structuredAttrs in setup hooks (part 2)

wolfgangwalther opened the pull request:
## Description of changes

This is a follow up to #318614, which had a lot more commits initially. After merging the basic changes for stdenv's appendTo, prependTo and new concatTo functions there, I now come back to apply those in more setup hooks.

I intentionally only took a part of long tail of commits I still have for this, so there will be more after this PR. All commits in this PR have one thing in common: I was able to test each of them by building the respective package mentioned in the commit message with `__structuredAttrs` turned on, similar to how I did it in #318614: The build for this package would fail on master, but succeed on this branch.

For some of the other setup hooks it's not that easy to find examples in nixpkgs, which actually fail with structuredAttrs turned on, so hard to prove the fix works. Thus I left them out for the moment and will bring them back once we have merged this set of changes, increasing the confidence in the general pattern of how the transition to `concatTo` works.

CC @emilazy @philiptaron @SomeoneSerge @tie 

Since the changed setup hooks won't trigger ofborg to mention the maintainers:
- bmake: @thoughtpolice @AndersonTorres 
- just: @xrelkd @06kellyjac 
- waf: @AndersonTorres 
- zig: @andrewrk @figsoda 
- build2: @vale981 @r-burns 
- scons: @AndersonTorres 
- premake: @bjornfor 
- gn: @stesie @matthewbauer @primeos 
- qt5: @qknight @ttuegel @periklis @bkchr 
- qt6: @milahu @NickCao 
- perl: @edolstra 

Quick recap for those not involved in #318614: That PR added a new function `concatTo` to allow merging of flags arrays / lists in either string or bash array format. Together with the already existing `appendTo` and `prependTo`, this allows to transparently support both building with and without `structuredAttrs` turned on, **without** putting checks whether structuredAttrs are enabled or not all over nixpkgs.

Making all the setup hooks support structuredAttrs is the first step to making this the default eventually.

## Things done

Of course I built the respective packages themselves - but since the setup hook is not a part of the derivation, those still work. As mentioned above, I built another package for each of the setup hooks, which uses that hook in the build process.

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wolfgangwalther committed:
waf: support structuredAttrs in setup hook

Tested samba4 with and without __structuredAttrs.

wolfgangwalther committed:
bmake: support structuredAttrs in setup hook

Tested libfsm with and without __structuredAttrs.

wolfgangwalther committed:
just: support structuredAttrs in setup hook

Tested cosmic-applibrary with and without __structuredAttrs.

wolfgangwalther committed:
zig: support structuredAttrs in setup hook

Tested waylock with and without __structuredAttrs.

wolfgangwalther committed:
build2: support structuredAttrs in setup hook

Tested build2 itself with and without __structuredAttrs.

wolfgangwalther committed:
scons: support structuredAttrs in setup hook

Tested swiften with and without __structuredAttrs.

wolfgangwalther committed:
scons: add missing checkTarget in setup hook

The checkTarget variable is explicitly tested before, but not passed to
scons.

This was apparently missed in the initial commit for that setupHook in
#50293.

wolfgangwalther committed:
premake: support structuredAttrs in setup hook

Tested gmad with and without __structuredAttrs.

Building it succeeded on master with structuredAttrs, but the log output
showed that the configure flags were not passed correctly previously.

wolfgangwalther committed:
just: shellcheck setup hook

wolfgangwalther committed:
build2: shellcheck setup hook

wolfgangwalther committed:
premake: shellcheck setup hook

wolfgangwalther committed:
gn: support structuredAttrs in setup hook

Tested v8 with and without __structuredAttrs.

wolfgangwalther committed:
gn: shellcheck setup hook

wolfgangwalther committed:
qt: support structuredAttrs in qmake hook

Tested qbs with and without __structuredAttrs.

wolfgangwalther committed:
perl: support structuredAttrs in generic builder

Tested perlPackages.GSSAPI with and without __structuredAttrs.

philiptaron commented on pkgs/development/tools/build-managers/build2/setup-hook.sh:
```diff
@@ -1,3 +1,5 @@
+# shellcheck shell=bash disable=SC2034,SC2157
```
## [SC2034](https://www.shellcheck.net/wiki/SC2034) foo appears unused. Verify it or export it.

What triggers this lint?

## [SC2157](https://www.shellcheck.net/wiki/SC2157) Argument to implicit -n is always true due to literal strings.

What triggers this lint?

philiptaron commented on pkgs/development/tools/build-managers/gn/setup-hook.sh:
```diff
@@ -1,14 +1,19 @@
+# shellcheck shell=bash disable=SC2164
```
## [SC2164](https://www.shellcheck.net/wiki/SC2164) Use `cd ... || exit` in case `cd` fails.

I'm assuming this is triggered by line 12. Could we land the disable on that line?

philiptaron commented on pkgs/development/libraries/qt-5/hooks/qmake-hook.sh:
```diff
@@ -1,40 +1,36 @@
+      "PREFIX=$out" \
+      "NIX_OUTPUT_OUT=$out" \
+      "NIX_OUTPUT_DEV=${!outputDev}" \
+      "NIX_OUTPUT_BIN=${!outputBin}" \
+      "NIX_OUTPUT_DOC=${!outputDev}/${qtDocPrefix:?}" \
+      "NIX_OUTPUT_QML=${!outputBin}/${qtQmlPrefix:?}" \
+      "NIX_OUTPUT_PLUGIN=${!outputBin}/${qtPluginPrefix:?}"
 
     if [ -n "@debug@" ]; then
-        qmakeFlags+=( "CONFIG+=debug" )
+        prependToVar qmakeFlags "CONFIG+=debug"
     else
-        qmakeFlags+=( "CONFIG+=release" )
+        prependToVar qmakeFlags "CONFIG+=release"
     fi
 
     # do the stripping ourselves (needed for separateDebugInfo)
-    qmakeFlags+=( "CONFIG+=nostrip" )
-
-    qmakeFlags+=( "${qmakeFlags_orig[@]}" )
+    prependToVar qmakeFlags "CONFIG+=nostrip"
```
Because these are multiple `prependToVar` calls, I think you need to reverse the order of the prepend calls: `CONFIG+=nostrip` first, followed by `CONFIG+=debug` or release, then finally the other ones that are first in this diff.

Maybe it doesn't matter though.

philiptaron commented on pkgs/development/compilers/zig/setup-hook.sh:
```diff
@@ -1,5 +1,6 @@
 # shellcheck shell=bash disable=SC2154,SC2086
```
## [SC2154](https://www.shellcheck.net/wiki/SC2154) var is referenced but not assigned.

What triggers this lint?

## [SC2086](https://www.shellcheck.net/wiki/SC2086) Double quote to prevent globbing and word splitting.

What triggers this lint?

philiptaron commented on pkgs/development/compilers/zig/setup-hook.sh:
```diff
@@ -1,5 +1,6 @@
 # shellcheck shell=bash disable=SC2154,SC2086
 
+# shellcheck disable=SC2034
```
## [SC2034](https://www.shellcheck.net/wiki/SC2034) foo appears unused. Verify it or export it.

Why does `zigDefaultFlagsArray` appear unused to shellcheck? It certainly _looks_ used below...

philiptaron commented on pkgs/development/compilers/zig/0.9/setup-hook.sh:
```diff
@@ -1,5 +1,6 @@
 # shellcheck shell=bash disable=SC2154,SC2086
 
+# shellcheck disable=SC2034
```
## [SC2034](https://www.shellcheck.net/wiki/SC2034) foo appears unused. Verify it or export it.

Why does `zigDefaultFlagsArray` appear unused to shellcheck? It certainly _looks_ used below...

philiptaron commented on pkgs/development/compilers/zig/0.9/setup-hook.sh:
```diff
@@ -1,5 +1,6 @@
 # shellcheck shell=bash disable=SC2154,SC2086
```
## [SC2154](https://www.shellcheck.net/wiki/SC2154) var is referenced but not assigned.

What triggers this lint?

## [SC2086](https://www.shellcheck.net/wiki/SC2086) Double quote to prevent globbing and word splitting.

What triggers this lint?

philiptaron commented on pkgs/by-name/ju/just/setup-hook.sh:
```diff
@@ -1,7 +1,10 @@
+# shellcheck shell=bash
```
Love to see it (especially with no lints disabled)

tie replied:
Out of curiosity, do we run shellcheck on CI for validation or this line is for manual editing?

philiptaron reviewed (commented):
Haven't tested it yet. Lots of questions about shellcheck; thanks for adding it in a lot of areas. The overall approach is sound.

AndersonTorres commented on pkgs/by-name/bm/bmake/setup-hook.sh:
```diff
@@ -85,9 +80,9 @@ bmakeDistPhase() {
     fi
 
     # Old bash empty array hack
```
Well I think this comment became outdated

wolfgangwalther committed:
zig: improve shellcheck disable comments

tie commented on pkgs/by-name/bm/bmake/setup-hook.sh:
```diff
@@ -42,11 +41,9 @@ bmakeCheckPhase() {
-            ${checkFlags:-VERBOSE=y} ${checkFlagsArray+"${checkFlagsArray[@]}"}
-            ${checkTarget}
         )
+        : "${checkFlags:=VERBOSE=y}"
```
I don’t think this is affects anything, but we did not assign default value for `checkFlags` globally before this change.

I’m not sure if this can be handled elegantly or whether if it has to be addressed at all. Perhaps
```bash
if [[ -v checkFlags ]]; then
    concatTo flagsArray checkFlags
else
    appendToVar flagsArray VERBOSE=y
fi
```
Same for other places where `${var:-default}` expansion was replaced with `${var:=default}` assignment.

Or set the default from `mkDerivation`.

>[!NOTE]
>Open for discussion, not a suggestion for change, not blocking.

emilazy replied:
I prefer the explicit helper.

tie replied:
I’m fine with either (as well as setting the default globally) since you’ve confirmed that nothing is affected by this change (https://github.com/NixOS/nixpkgs/pull/335666#discussion_r1722227995). Note that `local` variable is visible from child function calls, i.e. _post*_ hooks, but I don’t think that’s an issue.

tie reviewed (commented):
Looks good to me, although the changes introduce some side effects (see comment).

philiptaron reviewed (approved):
Thanks for addressing the shellcheck issues.

I like @tie's suggestion about 

```
if [[ -v checkFlags ]]; then
  concatTo flagsArray checkFlags
else
  appendToVar flagsArray VERBOSE=y
fi
```

philiptaron commented:
Is there anything that blocks this from being merged?

I move we merge it, and include it in the [new staging-next](https://github.com/NixOS/nixpkgs/pull/336718).

emilazy commented:
We can’t merge a nice‐to‐have Darwin stdenv rebuild into `staging-next`. Hydra has already built over 10,000 packages. It will have to go in the next cycle.

SomeoneSerge commented:
> `flagsArray` ... visible in `postX`-hooks

I have some concerns over what the `flagsArray` pattern means for https://github.com/NixOS/nixpkgs/issues/296544 (example: use part of the cmake build phase and part of the pypa-build phase in the same derivation)

wolfgangwalther committed:
stdenv: support default values in concatTo

The previously used pattern was introduced in #318614, but technically
leaked the default flags into the global scope. While this would
probably not make much of a practical difference, making concatTo
support default values is a much cleaner approach.

philiptaron commented on pkgs/stdenv/generic/setup.sh:
```diff
@@ -387,11 +387,15 @@ appendToVar() {
-        if type=$(declare -p "$name" 2> /dev/null); then
+    local arg default name type
+    for arg in "$@"; do
+        IFS="=" read -r name default <<< "$arg"
```
Could you add a test for a default that includes the `=` symbol?

tie commented:
Uh, yes, sorry, I completely forgot about this quirk when I’ve suggested using `[[ -v var ]]` test. See also https://stackoverflow.com/a/58261136
Perhaps we can check whether variable is defined with `declare -p var` (i.e. add an `else` branch for the check that we already have).

philiptaron reviewed (approved):
This all looks good to me.

tie commented:
>```
>IFS="=" read -r name default <<< "$arg"
>local -n nameref="$name"
>if [[ ! -n "${nameref[@]}" && -n "$default" ]]; then
>            targetref+=( "$default" )
>```
>@tie do you see any problem with that approach?

It treats an array with a single element that is an empty string as empty 😅
Hopefully not an issue in practice though.

<details>

```bash
declare -a arr=( '' )
if [[ -v arr ]]; then
  echo defined
fi
if [[ -z "${arr[@]}" ]]; then
  echo oops
fi
# Output:
# defined
# oops
```
</details>

What do you think about the following check?

```bash
# Either variable is not declared,
! type=$(declare -p -- "$name" 2>/dev/null) ||
  # or it is declared but no value is assigned (e.g. empty array),
  [[ -z ${!name+x} ]] ||
  # or it is not an array and is empty.
  [[ ${type#* } != -[aA]* && -z ${!name} ]]
```

That is, we’d treat empty values (empty arrays and strings) consistently.

For example,

<details>

```bash
set -eu -o pipefail

concatTo() {
    local -n targetref=$1
    shift
    local arg default name type
    for arg; do
        IFS='=' read -r name default <<<"$arg"
        # Append the default value if $name variable is not declared. Note that
        # variable attributes are separate from variable assignment. In addition
        # to that, declare assigns an attribute for name but does not actually
        # create the variable for empty arrays (i.e. `declare -a arr=()`).
        #
        # For compatibility, we treat variables with empty values as undeclared as well.
        if
            # Either variable is not declared,
            ! type=$(declare -p -- "$name" 2>/dev/null) ||
                # or it is declared but no value is assigned (e.g. empty array),
                [[ -z ${!name+x} ]] ||
                # or it is not an array and is empty.
                [[ ${type#* } != -[aA]* && -z ${!name} ]]
        then
            case "$default" in
            '('*)
                # Evaluate array expressions, i.e.
                # `concatTo fooFlags barFlags='(baz)'`.
                #
                # Beware that, just like the regular array assignments, this
                # evaluates Bash code inside the default value. E.g.
                # `declare -a arr='( "$(echo element)" )' && echo "${arr[*]}"` prints
                # `element`.
                local -a defaultArray=$default
                targetref+=("${defaultArray[@]}")
                ;;
            '')
                # Ignores both `undefinedVar=` and `undefinedVar` to avoid
                # appending an empty string for the latter.
                # TODO: perhaps return an error if no default is given and
                # variable is not declared?
                ;;
            *)
                targetref+=("$default")
                ;;
            esac
            continue
        fi

        local -n nameref=$name
        case "${type#* }" in
        -A*)
            echo "concatTo(): ERROR: trying to use concatTo on an associative array." >&2
            return 1
            ;;
        -a*)
            targetref+=("${nameref[@]}")
            ;;
        *)
            if [[ $name == *"Array" ]]; then
                targetref+=("${nameref[@]}")
            else
                # shellcheck disable=SC2206
                targetref+=(${nameref-})
            fi
            ;;
        esac
    done
}


declare -a definedFlags=(b c)
declare -a emptyArray=()
declare emptyString=
declare unboundString
declare string=h
concatTo flagsArray emptyArray=a definedFlags=unused undefinedWithDefaultArray='(d e)' emptyString=f unboundString=g string=unused undefinedWithoutDefault
echo "${flagsArray[@]}"
# Output:
# a b c d e f g h
```
</details>

Note that the code above also allows using _arrays_ as default values, e.g.
```bash
concatToVar flagsArray defaultArray='( --foo --bar --baz )'
```

tie commented:
>Yes, that's on purpose to not introduce a breaking change.

Perhaps a check for `[[ -z ${nameref[@]:+x} ]]` is more clear then, I usually expect `${var[@]}` to expand to multiple words (which I don’t think is the case here).

```console
$ bash -e -u -c 'declare -a arr=(); echo "${arr[@]:+x}"'

$ bash -e -u -c 'declare -a arr=(""); echo "${arr[@]:+x}"'

$ bash -e -u -c 'declare -a arr=("" ""); echo "${arr[@]:+x}"'
x
$ bash -e -u -c 'declare -a arr=("" "brr"); echo "${arr[@]:+x}"'
x
$ bash -e -u -c 'declare -a arr=("brr"); echo "${arr[@]:+x}"'
x
$ bash -e -u -c 's=; echo "${s[@]:+x}"'

$ bash -e -u -c 's=brr; echo "${s[@]:+x}"'
x
```
